### PR TITLE
[core] Violation suppressors

### PR DESCRIPTION
**Pull request by oowekyala**
* Add the interface ViolationSuppressor. Implementations filter violations before they reach the report.
  * Reimplement the logic of the violationSuppress_ properties, the NOPMD comments and annotation suppression with a ViolationSuppressor each
  * So they're all treated all uniformly instead of part in the Report, part in RuleViolation#isSuppressed (with overlap)
* Remove RuleViolation#isSuppressed. Suppressed rule violations now are always represented by a SuppressedViolation.
* The SuppressedViolation stores the ViolationSuppressor that suppressed the violation, so that renderers can report it correctly. This was previously hardcoded in renderers as just being comment suppression vs. annotation suppression, though other types of suppression exist.
* AbstractRuleViolationFactory is not abstract anymore as it's a completely functional implementation. I renamed it to DefaultRuleViolationFactory and removed the implementing classes that added no logic at all.
* Similarly, some implementations of RuleViolation add no logic and are unnecessary (ApexRuleViolation).
* Remove Parser#getSuppressMap from the Parser interface. This was just weird and made the parser stateful. It's on RootNode temporarily, though I don't like that either and think we could improve it later. I marked it InternalApi.

See also https://github.com/pmd/pmd/issues/1962#issuecomment-522544099

#### TODO on master

- [x] Deprecate Parser#getSuppressMap (internal api) #2084 
- [x] Mark the implementations of RuleViolationFactory and RuleViolation as internal API. The only published API there should be the interface.

#### Future

I'm still not satisfied with the current amount of boilerplate. Violation suppressors are orthogonal from the violation *factory* (the thing creating the violation), it shouldn't be necessary for a language implementation to extend the class just to provide new suppressors. Similarly, you shouldn't have to extend RuleViolation to add new info to the violation (class name, method name, etc). Implementations don't provide any behavior, they just populate fields in the constructor. An interface ViolationDecorator + a map of properties could be used instead. All that is todo for the future and is not the topic of this PR. I'll describe that on the wiki instead.

### CODE REVIEW

File: pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftHandler.java [hunk 1]
```diff
@@ -19,11 +17,6 @@ public XPathHandler getXPathHandler() {
         return new DefaultASTXPathHandler();
```
Comment (adangel) [inline pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftHandler.java]: The getXPathHandler method can also be removed here

File: pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/VfHandler.java [hunk 2]
```diff
@@ -19,11 +17,6 @@ public XPathHandler getXPathHandler() {
         return new DefaultASTXPathHandler();
```
Comment (adangel) [inline pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/VfHandler.java]: getXPathHandler() can be removed

File: pmd-vm/src/main/java/net/sourceforge/pmd/lang/vm/VmHandler.java [hunk 3]
```diff
@@ -23,10 +21,6 @@ public XPathHandler getXPathHandler() {
         return new DefaultASTXPathHandler();
```
Comment (adangel) [inline pmd-vm/src/main/java/net/sourceforge/pmd/lang/vm/VmHandler.java]: getXPathHandler() can be removed

File: pmd-xml/src/main/java/net/sourceforge/pmd/lang/xml/XmlHandler.java [hunk 4]
```diff
@@ -22,11 +20,6 @@ public XPathHandler getXPathHandler() {
         return new DefaultASTXPathHandler();
```
Comment (adangel) [inline pmd-xml/src/main/java/net/sourceforge/pmd/lang/xml/XmlHandler.java]: getXPathHandler() can be removed

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AnnotationSuppressionUtil.java [hunk 5]
```diff
@@ -103,6 +103,7 @@ private static Annotatable getSuppressor(Node node) {
         }
     }
 
+    @Nullable
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AnnotationSuppressionUtil.java]: Very true, this is a leftover from a refactoring I reverted

File: pmd-java/src/main/ant/alljavacc.xml [hunk 6]
```diff
@@ -131,8 +131,14 @@
                  token="throw new Error"
                  value="throw new RuntimeException" />
         <replace file="${target-package-dir}/ParseException.java"
-                 token="extends Exception"
-                 value="extends net.sourceforge.pmd.lang.ast.ParseException" />
+                 token="public class ParseException extends Exception"
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-java/src/main/ant/alljavacc.xml]: I think for some reason JavaCC still generates ParseException as public so this fixes it.

**Comment by oowekyala** [comment]
@adangel 

> Is this PR for java-grammar or for pmd/7.0.x? I think, this is a language agnostic API change... so, you need to switch the branches for this PR.

You're right it is.. I need to rework the branch a bit because I built it up on top of #1927, though as seen in the Apex module it isn't really necessary

Also I messed up the branch and this contains changes making parser support classes package-private (the branch started about that, and I forgot to branch away when shifting focus). So maybe:
1. One PR for java-grammar about parser class visibility
2. One PR for pmd/7.0.x with the violation suppressors
3. Fix merge conflicts later when merging back 7.0.x into java-grammar (conflicts caused by #1927) 

> Then next question: Which part/division of the PMD API does this belong to? Not the one, the rule implementors use, but the one, that language implementors use.

I think that's true. In fact, I think we could package those special APIs into `impl` subpackages, for code organisation.

So, for a package `p` for a feature (eg rule violations):
* published API are in `p` -> eg the interface `RuleViolation`, `ViolationSuppressor`
  * This should be enough to understand what the feature does while browsing javadocs
* implementation classes that we publish and support for language implementations to use are in `p.impl` -> eg the DefaultRuleViolationFactory, the RuleViolationFactory interface (if we need it anyway)
* internal API, that `p.impl` is built upon, or that we don't want to publish yet are in `p.internal`

I think this packaging scheme would help keep root packages clean of implementation details and help navigate the API. This is what I've started doing when cleaning up the packaging of the antlr adapters in b51c02f93ba230e50be4f64b462a6860e9dd1bef (part of #1796), eg

https://github.com/pmd/pmd/blob/55fea6a18fb7bdaa4e1d250506975f6ce06f75db/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrBaseNode.java#L5

Wdyt?

> We could have similar to LanguageVersionHandler#getXPathHandler a method "getAdditionalSuppressors"
> Closing down e.g. the RuleViolationFactory would reduce the API surface that we need to maintain later on.

I think that would be better yes. There is no point in having language-specific violation classes either. This is what I hinted at in the "future" section of the PR description. I think the data holder described on the wiki could be used to make rule violation extensible by any language module, and allow us to remove those language specific methods on the RuleViolation interface.

Excerpt from [the wiki](https://github.com/pmd/pmd/wiki/PMD-7.0.0-API) because it's not in a section so I can't link:
> Consider adding something like `T getData(DataKey<T> key)` to the node interface, where `DataKey` just ensures the visible API is strongly typed
>    * In fact, the designer already has something like that: [DataHolder](https://github.com/pmd/pmd-designer/blob/master/src/main/java/net/sourceforge/pmd/util/fxdesigner/util/DataHolder.java). Nodes should most probably contain a DataHolder of the sort. 
>    * The current `getUserData` would map naturally to this new system. 
>    * This makes the AST "extensible" externally by our internal subsystems, and by external tools (eg designer) 

There's much more to be said about how to improve our extension model but it's not the main topic of the PR

> We currently report about suppressed violations, but we don't report about "unused suppression markers"

Good idea! This looks like some kind of "meta rule" actually. 
More generally, it would be nice to have a rule to detect unused `@SuppressWarnings`. Eg a `@SuppressWarnings("unchecked")` when a compiler wouldn't report any unchecked warning anyway.

**Comment by adangel** [comment]
> 1. One PR for java-grammar about parser class visibility
> 2. One PR for pmd/7.0.x with the violation suppressors
> 3. Fix merge conflicts later when merging back 7.0.x into java-grammar (conflicts caused by #1927)

Sounds good.

> I think this packaging scheme would help keep root packages clean of implementation details and help navigate the API. This is what I've started doing when cleaning up the packaging of the antlr adapters in [b51c02f](https://github.com/pmd/pmd/commit/b51c02f93ba230e50be4f64b462a6860e9dd1bef) (part of #1796), eg
> 
> https://github.com/pmd/pmd/blob/55fea6a18fb7bdaa4e1d250506975f6ce06f75db/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrBaseNode.java#L5
> 
> Wdyt?

Yes, this package scheme looks good. We need to define however, what `p` will be and how many we will have. Some API classes will belong into one package (like RuleViolations), but certainly used by other packages as well (like used by Renderers).
Oh, and we don't need to do this right now with this PR. I've just noticed here a new class in the main package....

> There is no point in having language-specific violation classes either. This is what I hinted at in the "future" section of the PR description. I think the data holder described on the wiki could be used to make rule violation extensible by any language module, and allow us to remove those language specific methods on the RuleViolation interface.

Yes, that could be possible. Although we create a "Meta-API" then between a language and a renderer...e.g. the keys need to be known by the renderer so that it can display the method name and so on. Since renderers are language independent, the DataKeys need to be as well. So we might end up "optional" common keys, and renderers can display sometimes more information from a rule violation and sometimes not (if the language doesn't provide the info).
But that's for another PR.

> There's much more to be said about how to improve our extension model but it's not the main topic of the PR

exactly

> 
> > We currently report about suppressed violations, but we don't report about "unused suppression markers"
>
> Good idea! This looks like some kind of "meta rule" actually.
> More generally, it would be nice to have a rule to detect unused `@SuppressWarnings`. Eg a `@SuppressWarnings("unchecked")` when a compiler wouldn't report any unchecked warning anyway.

I was thinking, whether this could be part of this ViolationSuppressor interface? Something like `getAllSuppressionMarkers()` for a file. So that this functionality needs to be provided as part of the language implementation. We would then _somehow_ figure out afterwards, which of the markers have been used...

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by adangel** [comment]
Maybe we should do the two deprecations tasks on master before merging this PR, otherwise we might forget them... otherwise I'm fine with merging it.

**Pull request merged**